### PR TITLE
tt: working with a set of instances

### LINKED ISSUE #79: tt cartridge: start, stop, status, clean
Continuation of work on issue #34 

Modify commands to support cartridge-cli application:
* start
* stop
* status
* clean

### PR DESCRIPTION
**Pull request by 0x501D**
This patch adds the ability to works with a set of instances.
For example, we want to launch two instances based on one init.lua
file and one connected instance based on the router.lua file.
In order to do this, we create a directory called demo with the content:

init.lua:

    local inst_name = os.getenv('TARANTOOL_INSTANCE_NAME')
    local app_name = os.getenv('TARANTOOL_APP_NAME')

    while true do
        if app_name ~= nil and inst_name ~= nil then
            print(app_name .. ":" .. inst_name)
        else
            print("unknown instance")
        end
        require("fiber").sleep(1)
    end

router.init.lua:

    local inst_name = os.getenv('TARANTOOL_INSTANCE_NAME')
    local app_name = os.getenv('TARANTOOL_APP_NAME')

    while true do
        print("custom init file...")
        if app_name ~= nil and inst_name ~= nil then
            print(app_name .. ":" .. inst_name)
        else
            print("unknown instance")
        end
        require("fiber").sleep(1)
    end

instances.yml (The dot and dash characters in instance names
are reserved for system use.):

    router:
    master:
    replica:

Now we can run all instances at once:

   $ tt start demo
   • Starting an instance [router]...
   • Starting an instance [master]...
   • Starting an instance [replica]...

Or just one of them:

   $ tt start demo:master
   • Starting an instance [master]...

Closes #79

### CODE REVIEW

File: cli/running/running.go [hunk 1]
```diff
@@ -43,58 +43,156 @@ type RunOpts struct {
...
+	if cliOpts.App != nil {
+		appDir = cliOpts.App.InstancesAvailable
+	}
+
 	if appDir == "" {
 		if appDir, err = os.Getwd(); err != nil {
-			return "", err
+			return nil, err
 		}
+	} else if appDir[0:1] != "/" {
```
Comment (psergee) [inline cli/running/running.go:59]: Does not filepath.IsAbs do the job?

File: cli/running/running.go [hunk 2]
```diff
@@ -43,58 +43,156 @@ type RunOpts struct {
...
 		}
-		appPath = appAbsPath
+		appPath = luaPath
 	} else if dirStatErr == nil && dirInfo.IsDir() {
-		appPath = path.Join(dirAbsPath, "init.lua")
-		if _, err = os.Stat(appPath); err != nil {
-			return "", err
+		// Search for instances.yml
+		instCfgPath := path.Join(dirPath, "instances.yml")
+		if _, err = os.Stat(instCfgPath); err == nil {
```
Comment (psergee) [inline cli/running/running.go:98]: Can we check for != nil here? And move "main" code out of this condition to avoid indentation for huge code block?
The same comment for the line above:
`else if dirStatErr == nil && dirInfo.IsDir()`

File: cli/running/running.go [hunk 3]
```diff
@@ -43,58 +43,156 @@ type RunOpts struct {
...
+			}
+			instParams := make(map[string]interface{})
+			if err = yaml.Unmarshal(ymlData, instParams); err != nil {
+				return nil, err
+			}
+			for inst, _ := range instParams {
+				// Cartridge application uses dot and dash sepatator for the application
+				// and instance name (dash for stateboard, dot for others).
+				instance := cmdcontext.RunningCtx{}
+				sepIdx := -1
```
Comment (psergee) [inline cli/running/running.go:120]: Pls, extract this "search separator" code to separate function. So, it can be unit-tested easily. And will make this code a little simpler.

File: cli/running/running.go [hunk 4]
```diff
@@ -225,6 +323,18 @@ func createPIDFile(pidFileName string) error {
 	//    user:   read/write
 	//    group:  read
 	//    others: read
+	pidAbsDir := filepath.Dir(pidFileName)
+	if _, err := os.Stat(pidAbsDir); err != nil {
+		if os.IsNotExist(err) {
+			err = os.MkdirAll(pidAbsDir, 0770)
```
Comment (psergee) [inline cli/running/running.go:329]: Don't we want to make default dir permissions a constant?

File: cli/running/running.go [hunk 5]
```diff
@@ -225,6 +323,18 @@ func createPIDFile(pidFileName string) error {
 	//    user:   read/write
 	//    group:  read
 	//    others: read
+	pidAbsDir := filepath.Dir(pidFileName)
```
Comment (psergee) [inline cli/running/running.go:326]: Looks like, the comment above must be moved to 337 line. Because it is about the file creation, not a directory.

File: cli/running/running.go [hunk 6]
```diff
@@ -239,92 +349,145 @@ func createPIDFile(pidFileName string) error {
 	return nil
 }
 
+func makePath(path string, basePath string, inst *cmdcontext.RunningCtx) string {
```
Comment (psergee) [inline cli/running/running.go:352]: Pls, add a comment.

**Review by psergee — CHANGES_REQUESTED** [review]

File: cli/cartridge/extra/003_fix_work_paths.patch [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+--- a/cli/project/project.go
++++ b/cli/project/project.go
+@@ -23,8 +23,17 @@ func FillCtx(ctx *context.Ctx) error {
+ 		if err != nil {
+ 			return fmt.Errorf("Failed to get current directory: %s", err)
+ 		}
++		ctx.Running.AppDir = filepath.Join(ctx.Running.AppDir, ctx.Project.Name)
+ 	}
+ 
++	instAvail := os.Getenv("TT_INST_AVAILABLE")
```
[the remark this answers is not in the document]
Comment (0x501D) [reply cli/cartridge/extra/003_fix_work_paths.patch:120]: It uses in two different apps across (tt and cartridge-cli) 
So we cannot use one constant

File: cli/cartridge/extra/004_fix_warning.patch [hunk 8]
```diff
@@ -0,0 +1,16 @@
+diff --git a/cli/running/process.go b/cli/running/process.go
+index 9b56d17..758fd40 100644
+--- a/cli/running/process.go
++++ b/cli/running/process.go
+@@ -136,9 +136,9 @@ func (process *Process) SetPidAndStatus() {
+ 		return
+ 	}
+ 
+-	if name != "tarantool" {
++	/*if name != "tarantool" {
```
[the remark this answers is not in the document]
Comment (0x501D) [reply cli/cartridge/extra/004_fix_warning.patch:10]: In a non-commit patch, I see no difference

File: cli/cartridge/extra/003_fix_work_paths.patch [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
++			controlSocketSuffix := ".control"
++
++			if info.Mode()&os.ModeSocket == os.ModeSocket {
++				file := filepath.Base(path)
++				if !strings.HasSuffix(file, controlSocketSuffix) {
++					return nil
++				}
++
++				parts := strings.Split(path, "/")
++				instanceSocketPath := filepath.Join(ctx.Running.RunDir, parts[len(parts)-2])
```
[the remark this answers is not in the document]
Comment (0x501D) [reply cli/cartridge/extra/003_fix_work_paths.patch:56]: I need to cut only parent folder here, filepath.Dir cannot do the same.

File: cli/cmd/clean.go [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+	"github.com/apex/log"
+	"github.com/spf13/cobra"
+	"github.com/tarantool/tt/cli/cmdcontext"
+	"github.com/tarantool/tt/cli/configure"
+	"github.com/tarantool/tt/cli/modules"
+	"github.com/tarantool/tt/cli/running"
+)
+
+// NewStopCmd creates stop command.
```
[the remark this answers is not in the document]
Comment (0x501D) [reply cli/cmd/clean.go:14]: oops)

[the remark this answers is not in the document]
Comment (0x501D) [reply cli/cmd/clean.go]: Fixed

File: cli/cmd/connect.go [hunk 11]
```diff
@@ -53,25 +51,17 @@ func resolveInstAddr(cmdCtx *cmdcontext.CmdCtx, cliOpts *config.CliOpts,
...
-	for _, file := range files {
-		if file.Name() == newArgs[0]+".pid" {
-			newArgs[0] = filepath.Join(runDir, newArgs[0]+".control")
-			break
-		}
+	if len(cmdCtx.Running) > 1 {
+		return newArgs, fmt.Errorf("specify instance name")
 	}
 
+	newArgs[0] = cmdCtx.Running[0].ConsoleSocket
```
[the remark this answers is not in the document]
Comment (0x501D) [reply cli/cmd/connect.go:63]: Yes
// FillCtx returns error if no instances found.

File: cli/cmd/start.go [hunk 12]
```diff
@@ -55,22 +55,31 @@ func internalStartModule(cmdCtx *cmdcontext.CmdCtx, args []string) error {
...
+			if err := wdCmd.Start(); err != nil {
+				return err
+			}
+		}
 		return nil
 	}
 
-	log.Info("Starting an instance...")
-	if err = running.Start(cmdCtx); err != nil {
+	if err := running.Start(cmdCtx, &cmdCtx.Running[0]); err != nil {
```
[the remark this answers is not in the document]
Comment (0x501D) [reply cli/cmd/start.go:81]: There are only one instance will be launched after tt forks (see --watchdog section).

File: cli/cmd/connect.go [hunk 13]
```diff
@@ -53,25 +51,17 @@ func resolveInstAddr(cmdCtx *cmdcontext.CmdCtx, cliOpts *config.CliOpts,
...
-	runDir := cliOpts.App.RunDir
-	if runDir == "" {
-		if runDir, err = os.Getwd(); err != nil {
-			return newArgs, err
-		}
-	}
-
-	files, err := ioutil.ReadDir(runDir)
-	if err != nil {
+	// FillCtx returns error is no instances found.
```
[the remark this answers is not in the document]
Comment (0x501D) [reply cli/cmd/connect.go]: fixed

File: cli/cartridge/extra/003_fix_work_paths.patch (hunk 7 above)
[the remark this answers is not in the document]
Comment (psergee) [reply cli/cartridge/extra/003_fix_work_paths.patch:120]: Looks like this patch is applicable for cartridge-cli only.

File: cli/cmd/clean.go [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+			return nil
+		})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func askConfirm(question string) (bool, error) {
```
Comment (psergee) [inline cli/cmd/clean.go:57]: It could be useful to have "ask for confirmation" functionality in common place like utils or cli-utils, so other command can use it.

File: cli/running/running.go [hunk 15]
```diff
@@ -239,92 +349,145 @@ func createPIDFile(pidFileName string) error {
...
+	if path == "" {
+		if inst.SingleApp {
+			return filepath.Join(basePath, inst.AppName)
+		} else {
+			res = filepath.Join(basePath, inst.AppName)
+			return filepath.Join(res, inst.InstName)
+		}
+	}
+
+	if path[0:1] == "/" {
```
Comment (psergee) [inline cli/running/running.go:364]: filepath.IsAbs?

File: cli/cmd/clean.go [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+			return true, nil
+		}
+
+		if resp == "n" || resp == "no" {
+			return false, nil
+		}
+	}
+}
+
+func clean(run *cmdcontext.RunningCtx) error {
```
Comment (psergee) [inline cli/cmd/clean.go:79]: Why is there no unit test for this func?

File: cli/cartridge/extra/003_fix_work_paths.patch (hunk 9 above)
[the remark this answers is not in the document]
Comment (psergee) [reply cli/cartridge/extra/003_fix_work_paths.patch:56]: But using filepath api is safer. For instance, it will process this path correctly:
```go
"/a/b/c////socket.aaa"
filepath.Base(filepath.Dir(filename)) // will give "c"
parts[len(parts)-2] // will give an empty string
```

File: cli/cmd/clean.go (hunk 16 above)
Comment (0x501D) [reply cli/cmd/clean.go:79]: Added

File: cli/cartridge/extra/004_fix_warning.patch (hunk 8 above)
[the remark this answers is not in the document]
Comment (0x501D) [reply cli/cartridge/extra/004_fix_warning.patch:10]: fixed

File: cli/cartridge/extra/003_fix_work_paths.patch (hunk 7 above)
[the remark this answers is not in the document]
Comment (0x501D) [reply cli/cartridge/extra/003_fix_work_paths.patch:120]: fixed

File: cli/cartridge/extra/003_fix_work_paths.patch (hunk 9 above)
[the remark this answers is not in the document]
Comment (0x501D) [reply cli/cartridge/extra/003_fix_work_paths.patch:56]: fixed

File: cli/cmdcontext/cmdcontext.go [hunk 17]
```diff
@@ -64,6 +68,8 @@ type RunningCtx struct {
 	Restartable bool
 	// Control UNIX socket for started instance.
 	ConsoleSocket string
+	// True if this is a single application (no instances.yml).
```
[the remark this answers is not in the document]
Comment (0x501D) [reply cli/cmdcontext/cmdcontext.go:71]: yep, fixed.

File: cli/running/running.go (hunk 1 above)
Comment (0x501D) [reply cli/running/running.go:59]: Yes, fixed

File: cli/running/running.go (hunk 2 above)
Comment (0x501D) [reply cli/running/running.go:98]: in this place the old logic was extended without cardinal changes

File: cli/running/running.go (hunk 3 above)
Comment (0x501D) [reply cli/running/running.go:120]: Done

File: cli/running/running.go (hunk 5 above)
Comment (0x501D) [reply cli/running/running.go:326]: fixed

File: cli/running/running.go (hunk 4 above)
Comment (0x501D) [reply cli/running/running.go:329]: done

File: cli/running/running.go (hunk 6 above)
Comment (0x501D) [reply cli/running/running.go:352]: Comment added.

File: cli/cmd/clean.go [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+			}
+		},
+	}
+
+	cleanCmd.Flags().BoolVarP(&forceRemove, "force", "f", false, "do not ask for confirmation")
+
+	return cleanCmd
+}
+
+func collectFiles(list *[]string, dirname string) error {
```
[the remark this answers is not in the document]
Comment (0x501D) [reply cli/cmd/clean.go:38]: Why not. API is simpler.

File: cli/cmd/clean.go (hunk 14 above)
Comment (0x501D) [reply cli/cmd/clean.go:57]: moved to utils.go

File: cli/cmd/clean.go [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+		if err != nil {
+			return err
+		}
+	}
+
+	if confirm || forceRemove {
+		for _, file := range removeList {
+			err = os.Remove(file)
+			if err != nil {
+				return err
```
[the remark this answers is not in the document]
Comment (0x501D) [reply cli/cmd/clean.go:114]: Yes, in this case something goes wrong and we should stop operation.

File: cli/cmd/start.go [hunk 20]
```diff
@@ -55,22 +55,31 @@ func internalStartModule(cmdCtx *cmdcontext.CmdCtx, args []string) error {
 		if err != nil {
 			return err
 		}
-		newArgs := append([]string{"start", "--watchdog"}, args...)
+		for _, run := range cmdCtx.Running {
+			log.Infof("Starting an instance [%s]...", run.InstName)
```
[the remark this answers is not in the document]
Comment (0x501D) [reply cli/cmd/start.go:59]: Yes, that's right, we inherited such naming from our ancestors)

File: cli/running/running.go (hunk 15 above)
Comment (0x501D) [reply cli/running/running.go:364]: fixed

File: cli/running/running.go [hunk 21]
```diff
@@ -43,58 +43,156 @@ type RunOpts struct {
 }
 
 // findAppFile searches of an application init file.
-func findAppFile(appName string, cliOpts *config.CliOpts) (string, error) {
+func collectInstances(appName string, cliOpts *config.CliOpts,
```
[the remark this answers is not in the document]
Comment (0x501D) [reply cli/running/running.go:46]: updated

File: cli/running/running.go [hunk 22]
```diff
@@ -356,18 +519,25 @@ func Stop(cmdCtx *cmdcontext.CmdCtx) error {
...
 }
 
 // Status returns the status of the Instance.
-func Status(cmdCtx *cmdcontext.CmdCtx) string {
-	pid, err := getPIDFromFile(cmdCtx.Running.PIDFile)
+func Status(cmdCtx *cmdcontext.CmdCtx, run *cmdcontext.RunningCtx) string {
+	pid, err := getPIDFromFile(run.PIDFile)
 	if err != nil {
-		return fmt.Sprintf(`NOT RUNNING. Can't get the PID of process: "%v".`, err)
+		return fmt.Sprintf(`NOT RUNNING.`)
```
[the remark this answers is not in the document]
Comment (0x501D) [reply cli/running/running.go:540]: done

File: cli/cmd/clean.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (LeonidVas) [reply cli/cmd/clean.go:37]: I'm not very good at Go and can't tell if this is an idiomatic approach or not (and I don't think I'm the only one). I think it would be nice to add a link to the source of idiomatic (link to examples in popular repositories, code from a standard library, ...).

[the remark this answers is not in the document]
Comment (0x501D) [reply cli/cmd/clean.go]: fixed

**Review by psergee — APPROVED** [review]

File: .github/workflows/tests.yml [hunk 23]
```diff
@@ -81,6 +81,10 @@ jobs:
       - name: Unit tests
         run: mage unit
 
+      # This server starts and listen on 8084 port that is used for tests
```
Comment (LeonidVas) [inline .github/workflows/tests.yml:84]: Set a dot at the end of the sentence.

File: cli/running/running.go [hunk 24]
```diff
@@ -124,62 +132,163 @@ func (provider *providerImpl) IsRestartable() (bool, error) {
...
-		appPath = appAbsPath
+		appPath = luaPath
 	} else if dirStatErr == nil && dirInfo.IsDir() {
-		appPath = path.Join(dirAbsPath, "init.lua")
-		if _, err = os.Stat(appPath); err != nil {
-			return "", err
+		// Search for instances.yml
+		instCfgPath := path.Join(dirPath, "instances.yml")
+		if _, err = os.Stat(instCfgPath); err == nil {
+			// Get instances.
```
Comment (LeonidVas) [inline cli/running/running.go:212]: What do you think about moving the material to get the ctx array for multiinnstace into a separate function? Now it's hard to read.

File: cli/running/running.go [hunk 25]
```diff
@@ -209,12 +318,12 @@ func getPIDFromFile(pidFileName string) (int, error) {
 }
 
 // createLogger prepares a logger for the watchdog and instance.
-func createLogger(cmdCtx *cmdcontext.CmdCtx) *ttlog.Logger {
+func createLogger(cmdCtx *cmdcontext.CmdCtx, run *cmdcontext.RunningCtx) *ttlog.Logger {
```
Comment (LeonidVas) [inline cli/running/running.go:321]: Looks like `cmdCtx` is now unnecessary.

File: cli/running/running.go [hunk 26]
```diff
@@ -324,88 +445,149 @@ func createPIDFile(pidFileName string) error {
...
+		return filepath.Join(res, inst.InstName)
+	}
+
+	return res
+}
+
 // FillCtx fills the RunningCtx context.
 func FillCtx(cliOpts *config.CliOpts, cmdCtx *cmdcontext.CmdCtx,
 	args []string) error {
+	var err error
```
Comment (LeonidVas) [inline cli/running/running.go:492]: What for? The error is created as a local variable everywhere in this function.

File: test/utils.py [hunk 27]
```diff
@@ -65,10 +65,14 @@ def wait_file(dir_name, file_pattern, exclude_list, timeout_sec=1):
     iter_count = 0
 
     while True:
-        files = os.listdir(dir_name)
-        for file in files:
-            if re.match(file_pattern, file) is not None and file not in exclude_list:
-                return file
+        try:
```
Comment (LeonidVas) [inline test/utils.py:68]: It looks like it could be moved to a separate commit with an explanation of the changes in the commit message.

File: cli/cmd/root.go [hunk 28]
```diff
@@ -106,6 +106,7 @@ func NewCmdRoot() *cobra.Command {
 		NewPlayCmd(),
 		NewCartridgeCmd(),
 		NewCoredumpCmd(),
+		NewCleanCmd(),
```
Comment (LeonidVas) [inline cli/cmd/root.go:109]: Insert the new command at the end of the list.

**Review by LeonidVas — CHANGES_REQUESTED** [review]
Hi! Thank you for the patch.
See comments bellow.

File: .github/workflows/tests.yml [hunk 29]
```diff
@@ -165,5 +169,9 @@ jobs:
...
```
[the remark this answers is not in the document]
Comment (0x501D) [reply .github/workflows/tests.yml:172]: Fixed

File: .github/workflows/tests.yml (hunk 23 above)
[the remark this answers is not in the document]
Comment (0x501D) [reply .github/workflows/tests.yml:84]: Moved.

File: README.rst [hunk 30]
```diff
@@ -180,3 +180,70 @@ Common description. For a detailed description, use ``tt help command`` .
...
+-------------------------------
+
+``tt`` can manage a set of instances.
+For example, we want to launch two instances based on one ``init.lua`` file and one
+connected instance based on the ``router.lua`` file. In order to do this, we create
+a directory called ``demo`` with the content:
+
+``init.lua``:
+
+.. code-block:: lua
```
[the remark this answers is not in the document]
Comment (0x501D) [reply README.rst]: doc/examples.rst was created.

File: cli/cmd/check.go [hunk 31]
```diff
@@ -37,10 +37,28 @@ func internalCheckModule(cmdCtx *cmdcontext.CmdCtx, args []string) error {
 		return err
 	}
 
-	if err := running.Check(cmdCtx); err != nil {
-		return err
+	// Сollect a list of instances with unique scripts.
```
[the remark this answers is not in the document]
Comment (0x501D) [reply cli/cmd/check.go:40]: Cannot move this to a separate commit, tt will not compile.

File: cli/cmd/logrotate.go [hunk 32]
```diff
@@ -13,7 +13,7 @@ import (
 func NewLogrotateCmd() *cobra.Command {
 	var logrotateCmd = &cobra.Command{
 		Use:   "logrotate <INSTANCE_NAME>",
-		Short: "Rotate logs of a started tarantool instance",
+		Short: "Rotate logs of a started tarantool instance(s)",
```
[the remark this answers is not in the document]
Comment (0x501D) [reply cli/cmd/logrotate.go:16]: Cannot move this to a separate commit, tt will not compile. (same as check)

File: cli/cmd/connect.go [hunk 33]
```diff
@@ -53,25 +51,17 @@ func resolveInstAddr(cmdCtx *cmdcontext.CmdCtx, cliOpts *config.CliOpts,
...
-	for _, file := range files {
-		if file.Name() == newArgs[0]+".pid" {
-			newArgs[0] = filepath.Join(runDir, newArgs[0]+".control")
-			break
-		}
+	if len(cmdCtx.Running) > 1 {
+		return newArgs, fmt.Errorf("specify instance name")
 	}
 
+	newArgs[0] = cmdCtx.Running[0].ConsoleSocket
```
[the remark this answers is not in the document]
Comment (0x501D) [reply cli/cmd/connect.go:63]: `tt connect` can connect only to one instance (as design). So user must specify instance if set ot instances used: example
`tt connect demo:inst0`. If running.FillCtx() collects more than one instance, error will be raised: `specify instance name`

File: cli/cmd/root.go [hunk 34]
```diff
@@ -145,6 +145,10 @@ func InitRoot() {
 		log.Fatalf("Failed to get Tarantool CLI configuration: %s", err)
 	}
 
+	if cliOpts.App != nil {
```
[the remark this answers is not in the document]
Comment (0x501D) [reply cli/cmd/root.go:149]: Done

File: cli/cmd/start.go [hunk 35]
```diff
@@ -55,22 +55,31 @@ func internalStartModule(cmdCtx *cmdcontext.CmdCtx, args []string) error {
...
+		}
 		return nil
 	}
 
-	log.Info("Starting an instance...")
-	if err = running.Start(cmdCtx); err != nil {
+	if err := running.Start(cmdCtx, &cmdCtx.Running[0]); err != nil {
 		return err
 	}
-
```
[the remark this answers is not in the document]
Comment (0x501D) [reply cli/cmd/start.go:74]: fixed

File: cli/running/instance.go [hunk 36]
```diff
@@ -22,7 +23,9 @@ type Instance struct {
 	// that will be used to launch the Instance.
 	tarantoolPath string
 	// appPath describes the path to the "init" file of an application.
-	appPath string
+	appPath  string
+	appName  string
```
[the remark this answers is not in the document]
Comment (0x501D) [reply cli/running/instance.go:27]: Description added.

File: cli/running/running.go [hunk 37]
```diff
@@ -18,7 +18,14 @@ import (
 	"github.com/tarantool/tt/cli/config"
 	"github.com/tarantool/tt/cli/configure"
 	"github.com/tarantool/tt/cli/ttlog"
-	"github.com/tarantool/tt/cli/util"
+	"gopkg.in/yaml.v2"
+)
+
+const (
+	defaultDirPerms  = 0770
+	InstStateStopped = "NOT RUNNING."
```
[the remark this answers is not in the document]
Comment (0x501D) [reply cli/running/running.go:26]: Moved.

File: cli/running/running.go [hunk 38]
```diff
@@ -45,7 +52,8 @@ type RunOpts struct {
 
 // providerImpl is an implementation of Provider interface.
 type providerImpl struct {
-	cmdCtx *cmdcontext.CmdCtx
+	cmdCtx   *cmdcontext.CmdCtx
+	instance *cmdcontext.RunningCtx
```
[the remark this answers is not in the document]
Comment (0x501D) [reply cli/running/running.go:56]: Comment added.

File: cli/running/running.go (hunk 24 above)
Comment (0x501D) [reply cli/running/running.go]: Done

File: cli/running/running.go (hunk 26 above)
Comment (0x501D) [reply cli/running/running.go:493]: This needs to prevent shadowing `basePath` in this part:
```go
    if basePath == "" {                                                                                   
        if basePath, err = os.Getwd(); err != nil {                                                       
            return fmt.Errorf(`Can't get the "RunDir: %s"`, err)                                          
        }                                                                                                 
    } 
    ```

File: test/utils.py (hunk 27 above)
Comment (0x501D) [reply test/utils.py:68]: Moved to separate commit.

File: cli/cmd/root.go (hunk 28 above)
Comment (0x501D) [reply cli/cmd/root.go:109]: fixed

**Review by LeonidVas — APPROVED** [review]
LGTM
